optaplanner-core/src/test/java/org/optaplanner/core/api/solver/SolverManagerTest.java

```diff
@@ -240,4 +255,170 @@ public void skipAhead() throws ExecutionException, InterruptedException {
+        assertSolveWithoutConsumer(problemCount, solverManager);
+        assertSolveWithConsumer(problemCount, solverManager, true);
+        assertSolveWithConsumer(problemCount, solverManager, false);
+    }
+
+    private void assertSolveWithoutConsumer(int problemCount, SolverManager<TestdataSolution, Integer> solverManager)
+            throws InterruptedException, ExecutionException {
+        List<SolverJob<TestdataSolution, Integer>> jobs = new ArrayList<>(problemCount);
+
+        for (int id = 0; id < problemCount; id++) {
+            jobs.add(solverManager.solve(id, PlannerTestUtils.generateTestdataSolution(Integer.toString(id))));
+        }
+        assertInitializedJobs(jobs);
+    }
+
+    private void assertSolveWithConsumer(
+            int problemCount, SolverManager<TestdataSolution, Integer> solverManager, boolean onlyFinalBestSolutions)
+            throws ExecutionException, InterruptedException {
+
+        // Two solutions for every problem.
+        Map<Integer, List<TestdataSolution>> consumedSolutions = new HashMap<>(problemCount * 2);
+
+        List<SolverJob<TestdataSolution, Integer>> jobs = new ArrayList<>(problemCount);
+
+        for (int id = 0; id < problemCount; id++) {
+            List<TestdataSolution> consumedBestSolutions = Collections.synchronizedList(new ArrayList<>());
+            String solutionName = Integer.toString(id);
+            if (onlyFinalBestSolutions) {
+                jobs.add(solverManager.solve(
+                        id,
+                        problemId -> PlannerTestUtils.generateTestdataSolution(solutionName, 2),
+                        consumedBestSolutions::add, null));
+            } else {
+                jobs.add(solverManager.solveAndListen(
+                        id,
+                        problemId -> PlannerTestUtils.generateTestdataSolution(solutionName, 2),
+                        consumedBestSolutions::add, null));
+            }
+            consumedSolutions.put(id, consumedBestSolutions);
+        }
+        assertInitializedJobs(jobs);
+        assertConsumedSolutions(problemCount, onlyFinalBestSolutions, consumedSolutions);
+    }
+
+    private void assertConsumedSolutions(int problemCount, boolean onlyFinalBestSolutions, Map<Integer, List<TestdataSolution>> consumedSolutions) {
+        if (onlyFinalBestSolutions) {
+            assertConsumedFinalBestSolutions(problemCount, consumedSolutions);
+        } else {
+            assertConsumedBestSolutions(problemCount, consumedSolutions);
+        }
+    }
+
+    private void assertConsumedFinalBestSolutions(int problemCount, Map<Integer, List<TestdataSolution>> consumedSolutions) {
+        for (int i = 0; i < problemCount; i++) {
+            assertThat(consumedSolutions.get(i)).hasSize(1);
+            TestdataSolution solution = consumedSolutions.get(i).get(0);
+            assertConsumedFinalBestSolution(solution);
+        }
```

Comment:
Sorry, I can't follow how the code flows here. It's not readable.
Try unextracting all methods starting with assertConsumed and then figuring out which methods make sense to extract. Avoid using both "assertConsumedFinalBestSolutions" and "assertConsumedFinalBestSolution"

Reply:
There is SOFT guideline to avoid having more than 20 lines in a single method. Don't hesitate to violate that guideline to keep the code readable, it's not a hard guideline. Especially for tests, 20 lines is awfully low IMO.